src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -520,7 +537,11 @@ private static IEnumerable<FunctionOverload> GetSystemOverloads(IFeatureProvider
                 .WithRequiredParameter("predicate", OneParamLambda(LanguageConstants.Any, LanguageConstants.Bool), "The predicate applied to each input array element. If false, the item will be filtered out of the output array.",
                     calculator: getArgumentType => CalculateLambdaFromArrayParam(getArgumentType, 0, t => OneParamLambda(t, LanguageConstants.Bool)))
                 .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) => {
-                    return new(argumentTypes[0]);
+                    return new(argumentTypes[0] switch
+                    {
+                        TupleType tuple => new TypedArrayType(tuple.Item, TypeSymbolValidationFlags.Default),
```

Comment:
Why do you need to transform the output value in this case?

Reply:
If a tuple gets filtered, you know that each member of the resulting array is assignable to a union of the types in the original tuple, but you can no longer make any assertions about specific indices. For example, in:

```bicep
var a = range(0, 5) // type<[0, 1, 2, 3, 4]>
var b = filter(a, x => x % 2 == 0) // type<(0 | 1 | 2 | 3 | 4)[]>
```

`b` still has the same `.Item` type signature as `a`, but running the value through a filter function means that it's not safe to use the same tuple type for the input array and return type of `filter`. In this case, the return type should be `[0, 2, 4]`, which is assignable to `type<(0 | 1 | 2 | 3 | 4)[]>` but not to `type<[0, 1, 2, 3, 4]>`

Reply:
Ah of course! Thanks for the explanation. Probably wouldn't hurt to add that as a comment to the code as well!